### DISCUSSION TITLE
[typebox/value] Property Transformation from One Property to Another

### DISCUSSION THREAD

**Discussion by moontai0724** [sinclairzx81/typebox#514] [Ideas]
Currently, the TypeBox provides useful functions like `Cast` in `Value` to transform an object into a target type. However, the transformation is only possible when the property matches the key.

I think to introducing a new feature that allows property transformation from one property to another would be highly beneficial and enhance the capabilities of TypeBox.

```typescript
import { Type } from '@sinclair/typebox';
import { Value } from '@sinclair/typebox/value';

const User = Type.Object(
  {
    id: Type.String({
      key: 'userId',
    }),
  },
);

const transformResult = Value.Cast(User, {
  userId: '100',
});

console.log(transformResult); // Expect: `{ id: '100' }`
```

**Comment by sinclairzx81** [comment]
@moontai0724 Hi,

There is a partially implemented version of transforms available in the `examples/transform` directory.

https://github.com/sinclairzx81/typebox/tree/master/examples/transform

```typescript
import { Transform, Encode, Decode } from './transform'

const User = Transform(Type.Object({
  id: Type.String()
}), {
  decode: (value) => {
    return { userId: value.id }
  },
  encode: (value) => {
    return { id: value.userId }
  }
})

const decoded = Decode(User, { id: '100' })  // const decoded = { userId: '100' }

const encoded = Encode(User, decoded)        // const encoded = { id: '100' }
```
This implementation is pending inclusion in the library until it's had a good review and all the requirements ironed out, however you can use this module today by copying and pasting the `transform.ts` into your project. All feedback is welcome on the module.

**Reply by moontai0724** [reply]
Excellent! looking forward to release!

**Reply by sinclairzx81** [reply]
@moontai0724 Hi,

There feature should be landing very soon. There is currently tagged `dev` revision published to NPM that you can install and try today.
```typescript
npm install @sinclair/typebox@dev
```
Revision notes for the release can be found on the PR here https://github.com/sinclairzx81/typebox/pull/525
Cheers
S
